### PR TITLE
many: ignore configure hook failures on core refresh

### PR DESCRIPTION
**Pull request by mvo5**
This branch makes snapd ignore (but log) configure hooks failures from the core snap on refresh. This ensures that we can always to a new core snap without being blocked on a bug in the configure hook handling.

If the approach taken looks reasonable I will:
a) add code that sends an error report about this to errors.ubuntu.com (probably in a folllowup branch though)

We should squash merge this to make cherry-picking easier.

### CODE REVIEW

File: overlord/hookstate/hookmgr.go [hunk 1]
```diff
@@ -216,8 +218,13 @@ func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
 
 	if hookExists {
 		output, err := runHook(context, tomb)
-		if err != nil {
+		if hooksup.IgnoreFail && err != nil {
+			task.State().Lock()
+			task.Logf("ignoring failure in hook %q: %v", hooksup.Hook, err)
+			task.State().Unlock()
+		} else if err != nil {
 			err = osutil.OutputErr(output, err)
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: you probably want to do this on both paths? no

**Comment by mvo5** [comment]
Closing for now, I will add the spread test right away because whatever the implementation that will be needed anyway.

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -225,14 +230,18 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 
 	installSet := state.NewTaskSet(tasks...)
 
-	configSet := Configure(st, snapsup.Name(), defaults)
+	var confFlags int
+	if snapsup.Name() == "core" {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: this bit probably merits a comment? checking on name feels always a bit strange, we don't have the type until we actually run the change though

Comment (mvo5) [inline overlord/snapstate/snapstate.go]: Added a comment now.

File: overlord/hookstate/hookmgr.go (hunk 1 above)
Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: You mean in this error path here? https://github.com/snapcore/snapd/pull/3071/files/a27f715bda5bc5cae149282695a92a6c126f0848#diff-87319759a99457f870b9a8a989aa7867L229 ? I can add it there, yes, but for our case it looks like the handlers for Done() set in configstate is always returning nil. I could rename "ignoreFail" to "ignoreHookFail" maybe? Or maybe I am missing a bit of the bigger picture here :)

Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: sorry, I mean do this before the if ? it seems we are not logging the output if ignore the failure

Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: Thanks a lot! this was indeed missing, I added it and updated the test.

File: overlord/snapstate/snapmgr_test.go [hunk 3]
```diff
@@ -863,6 +863,34 @@ func (s *snapmgrTestSuite) TestUpdateTasks(c *C) {
 	c.Check(snapsup.Channel, Equals, "some-channel")
 }
 
+func (s *snapmgrTestSuite) TestUpdateTasksCoreSetsIgnoreOnConfiugre(c *C) {
```
Comment (pedronis) [inline overlord/snapstate/snapmgr_test.go]: typo: Configure

File: tests/main/refresh-core-with-failing-configure-hook/task.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    echo "Breaking the configure hook of the core snap"
+    unsquashfs /var/lib/snapd/snaps/core_$(cat prevBoot).snap
+    printf '#!/bin/sh\necho somethingbroken\nfalse\n' > squashfs-root/meta/hooks/configure
+    chmod 755 squashfs-root/meta/hooks/configure
+
+    . $TESTSLIB/snaps.sh
+    mksnap "squashfs-root" "$BROKEN_CORE_SNAP"
+    rm -rf squshfs-root
+
+    echo "Installing a new core which will trigger running the conifgure hook"
```
Comment (pedronis) [inline tests/main/refresh-core-with-failing-configure-hook/task.yaml]: typo: configure

File: overlord/hookstate/hookmgr.go [hunk 5]
```diff
@@ -218,11 +220,17 @@ func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
 		output, err := runHook(context, tomb)
 		if err != nil {
 			err = osutil.OutputErr(output, err)
-			if handlerErr := context.Handler().Error(err); handlerErr != nil {
-				return handlerErr
+			if hooksup.IgnoreFail {
+				task.State().Lock()
+				task.Logf("ignoring failure in hook %q: %v", hooksup.Hook, err)
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: I think you can use Errof here, is a kind of error even if it doesn't make the task fail

File: overlord/snapstate/snapmgr_test.go (hunk 3 above)
Comment (mvo5) [inline overlord/snapstate/snapmgr_test.go]: Thanks, fixed.

File: tests/main/refresh-core-with-failing-configure-hook/task.yaml (hunk 4 above)
Comment (mvo5) [inline tests/main/refresh-core-with-failing-configure-hook/task.yaml]: Thanks! Even spelled different than above :/ - fixed now as well :)

File: overlord/hookstate/hookmgr.go (hunk 5 above)
Comment (mvo5) [inline overlord/hookstate/hookmgr.go]: Indeed, I added an task.Errorf() here now.

File: overlord/configstate/tasksets.go [hunk 6]
```diff
@@ -33,11 +33,12 @@ func init() {
...
-func Configure(s *state.State, snapName string, patch map[string]interface{}) *state.TaskSet {
+func Configure(s *state.State, snapName string, patch map[string]interface{}, flags int) *state.TaskSet {
 	hooksup := &hookstate.HookSetup{
-		Snap:     snapName,
-		Hook:     "configure",
-		Optional: len(patch) == 0,
+		Snap:       snapName,
+		Hook:       "configure",
+		Optional:   len(patch) == 0,
+		IgnoreFail: flags&snapstate.IgnoreHookFailure > 0,
```
Comment (zyga) [inline overlord/configstate/tasksets.go]: (nitpick land) The `> 0` part is a bit unusual, in C that would be sufficient by itself, perhaps we can just use `!= 0`

File: tests/lib/snaps.sh [hunk 7]
```diff
@@ -14,3 +14,15 @@ install_local() {
 install_local_devmode() {
     install_local "$1" --devmode
 }
+
+mksnap() {
```
Comment (zyga) [inline tests/lib/snaps.sh]: We already have functions for doing this, we could use those or special-case this for core (mk_core_snap) or something. Not urgent in this branch. Perhaps @fgimenez has an opinion.

Comment (pedronis) [inline tests/lib/snaps.sh]: bit confused, which functions? where?

**Pull request merged**